### ISSUE TITLE
Make tile cache size customizable to prevent iOS crashes

### ISSUE THREAD

**Issue by kristfal** [mapbox/mapbox-gl-js#4052]
Hey,

one of our core target devices are iPads with 1GB RAM and more. We're running a single MGL instance covering the full viewport. In that environment, we're experiencing a lot of crashes on iPads due to memory pressure, and some crashes on iPhones. 

The [sourceCache](https://github.com/mapbox/mapbox-gl-js/blob/67fc7dff69d512db92a30605ca533acfbb7100fb/js/source/source_cache.js#L279) assigns a cache size based on the viewport size of the device. 

Our current workaround is to reduce the cache size from the predefined 45 tiles to 8 on iPads. We do a similar but not as drastic reduction on iPhones. This workaround reduces memory related crashes quite a bit, but they are not fully sufficient to prevent a 1GB RAM iPad to handle zoom and pan interactions for more than a couple of minutes before crashing.

With that workaround, an emulated iPad in Chrome keeps a stable heap size of 60 - 70 MB in our test bench. 

In Mobile Safari, we see the test starting at 35 MB heap size, and then gradually growing up to 130+ MB (as the image bellow illustrates). Eventually either the inspector or safari crashes.

![skjermbilde 2017-01-26 17 48 04](https://cloud.githubusercontent.com/assets/5638107/22341265/746822b2-e3f0-11e6-809d-b0cead8d2c89.png)

**mapbox-gl-js version**: Master

### Steps to Trigger Behavior

 1. Run mapbox-gl-js in full screen on an iPad with 1 GB RAM
 2. Zoom and pan (for instance use flyTo) inside or between larger cities
 3. Crash

### Expected Behavior

iOS devices should not crash after prolonged use.

### Actual Behavior

Retained memory builds on iOS, triggering crashes after prolonged use.

### Suggestions

– Reduce cache size on iOS
– Identify potential memory leaks in Safari (let me know if any heap snapshots would help)

Side note: We are only able to record timelines in Safari when the map is not moving. The best approach seems to be to interact with the map for 30 seconds, pause and record with the inspector for 10 seconds and then repeat until either the inspector or safari crashes.

**Comment by stackTom** [comment]
I have a similar issue: https://github.com/mapbox/mapbox-gl-js/issues/4063

**Comment by kristfal** [comment]
@stackTom Try replacing `updateCacheSize` with:

```js
        const widthInTiles = Math.ceil(transform.width / transform.tileSize) + 1;
        const heightInTiles = Math.ceil(transform.height / transform.tileSize) + 1;
        const approxTilesInView = widthInTiles * heightInTiles;
        const commonZoomRange = 5;
        const maxSize = Math.floor(approxTilesInView * commonZoomRange);
        const ios = /iP(hone|od|ad)/.test(navigator.userAgent);

        // Constrain cache on iOS devices
        if (ios) {maxSize = 8;}

        this._cache.setMaxSize(maxSize);
```

On a side note: We're also seing frequent crashes on iPads with 2GB RAM too. I'm starting to think that an underlying WebGL issue could trigger crashes as well.

**Comment by stackTom** [comment]
@kristfal thanks, will try it out. Hope this issue gets fixed soon

**Comment by stackTom** [comment]
@kristfal I am assuming you installed with browserify so you can modify the code? Do you have a link on how to do this? I can't figure it out.

**Comment by kristfal** [comment]
@stackTom We build from source and bundle it with our own apps. I suggest you do the same. See [Contributing.md](https://github.com/mapbox/mapbox-gl-js/blob/master/CONTRIBUTING.md) for details on how to build the repo.

**Comment by stackTom** [comment]
@kristfal Thanks, it worked great. Hope it alleviates some of my crashes. Are there any significant disadvantages to reducing cache size (speed I'm assuming)?

**Comment by kristfal** [comment]
@stackTom there are a couple of downsides:

1) You'll notice more 'white areas' while waiting to download and parse tiles in areas that you've recently visited
2) Depending on your cache settings, you might use more bandwidth 
3) Mobile specific: Depending on the usage pattern, you may notice slightly more garbage collection freezes as you're throwing away parsed tiles more often and building more garbage.

All in all, this tweak is no magic bullet, but it helped us quite a bit.

**Comment by mourner** [comment]
Would it help to expose this cache number as a `Map` option instead of having to fork?

**Comment by NGMarmaduke** [comment]
@mourner It would, we forked in the interest of time

**Comment by mourner** [comment]
@NGMarmaduke want to submit a PR?

**Comment by andrewharvey** [comment]
I believe this issue "Make tile cache size customizable" is fixed since #4778 was merged, so I'll close this issue, please re-open if I'm missing something.
